In [261]:
import os
os.getcwd()

'D:\\'

In [269]:
os.chdir("D://")

In [333]:
import pandas as pd

df = pd.read_excel("para_data.xlsx")
df = df[['text', 'label']][:2166]

In [334]:
df

text  label
0     Today a variety of additive manufacturing (AM)...      0
1     The Ti-6Al-4V alloy, a popular α (hexagonal cl...      0
2     The microstructure transformation is another r...      0
3     For LMWD of Ti-6Al-4 V alloy, solutions for im...      0
4     Accordingly, this work aims to study the influ...      0
...                                                 ...    ...
2161  where α, h,υ, A, Eg and n: the absorption coef...      0
2162  UV–visible absorption spectra of ZnO:NiO nano-...      0
2163                             Energy gap measurement      0
2164  NiO:ZnO has an optical energy band gap of 3.57...      0
2165  The PLD method is an environmentally friendly,...      0

[2166 rows x 2 columns]

In [309]:
df[df['label'] == 1].shape[0]

59

In [310]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

In [311]:
X_train, X_test, y_train, y_test = train_test_split(
    df['text'], df['label'], test_size=0.20, stratify=df['label'], random_state=42
)

In [ ]:
vectorizer = TfidfVectorizer(max_features=10000, ngram_range=(1, 2))
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

# you change max features

In [ ]:

model = LogisticRegression(class_weight='balanced', max_iter=1000)
model.fit(X_train_vec, y_train)
lr_model=model

#iteration = 1000 for good results

In [314]:
y_pred = model.predict(X_test_vec)

print("Classification Report:")
print(classification_report(y_test, y_pred, digits=4))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0     0.9974    0.9871    0.9922       388
           1     0.6875    0.9167    0.7857        12

    accuracy                         0.9850       400
   macro avg     0.8424    0.9519    0.8890       400
weighted avg     0.9881    0.9850    0.9860       400

Confusion Matrix:
[[383   5]
 [  1  11]]


In [315]:
from sklearn.metrics import roc_auc_score


y_pred_prob = model.predict_proba(X_test_vec)[:, 1]
auc_roc = roc_auc_score(y_test, y_pred_prob)
print(f"AUC-ROC: {auc_roc}")

AUC-ROC: 0.9974226804123711


In [316]:

cm = confusion_matrix(y_test, y_pred)

TP = cm[1, 1] 
FP = cm[0, 1] 
TN = cm[0, 0] 
FN = cm[1, 0] 

print(f"True Positives (TP): {TP}")
print(f"False Positives (FP): {FP}")
print(f"True Negatives (TN): {TN}")
print(f"False Negatives (FN): {FN}")


True Positives (TP): 11
False Positives (FP): 5
True Negatives (TN): 383
False Negatives (FN): 1


# tried different models, but worst result.

In [ ]:
#worse results
#try svm, or other

# again worst results, except svm
# stick to logistic regression only with tf-idf

# Tweaking and Oversampling

In [259]:
import json
import pandas as pd


with open('paragraphs.json', 'r', encoding='utf-8') as file:
    data = json.load(file)

rows = [{'Paragraph Number': int(k), 'Paragraph': v['paragraph']} for k, v in data.items()]

df = pd.DataFrame(rows).sort_values('Paragraph Number')

df.to_excel('paragraphsbefore.xlsx', index=False, engine='openpyxl')


In [260]:
df


Paragraph Number                                          Paragraph
0                    0  The SLs were grown by pulsed laser deposition ...
1                    1  Thin films of w-BN are grown by pulsed laser d...
2                    2  Interconnected ZnO core/shell nanorods were gr...
3                    3  Before deposition, the substrates were heated ...
4                    4  SrHA nanocrystals were synthesized in N2 atmos...
...                ...                                                ...
8806              8806  The base pressure inside the deposition chambe...
8807              8807  Pulsed laser deposition technique was used to ...
8808              8808  SrTiO3-δ homoepitaxial thin films were grown u...
8809              8809  The growth sequence is 3 × (5 unit cells of Sr...
8810              8810  A new series of rare earth-niobate, RE3NbO7 (R...

[8811 rows x 2 columns]

In [335]:
import pandas as pd

df2 = pd.read_excel("paragraphsextra.xlsx")
df2 = df2[['Paragraph', 'Label']][:76]


In [336]:
df2 = df2.rename(columns={
    'Paragraph': 'text',
    'Label': 'label'
})

df2

text  label
0   The SLs were grown by pulsed laser deposition ...    1.0
1   Thin films of w-BN are grown by pulsed laser d...    1.0
2   Interconnected ZnO core/shell nanorods were gr...    1.0
3   Before deposition, the substrates were heated ...    1.0
4   SrHA nanocrystals were synthesized in N2 atmos...    0.0
..                                                ...    ...
71  The TiON nanofilms were prepared by pulsed las...    1.0
72  Samples were grown starting from commercially ...    1.0
73  Circular pellets of diameter 10 mm were made a...    1.0
74  The MSPLD system scheme is presented in Fig. 1...    1.0
75  Commercially available targets of Al2O3, BaF2,...    1.0

[76 rows x 2 columns]

In [337]:
import pandas as pd

merged_df = pd.concat([df, df2], ignore_index=True)

merged_df

text  label
0     Today a variety of additive manufacturing (AM)...    0.0
1     The Ti-6Al-4V alloy, a popular α (hexagonal cl...    0.0
2     The microstructure transformation is another r...    0.0
3     For LMWD of Ti-6Al-4 V alloy, solutions for im...    0.0
4     Accordingly, this work aims to study the influ...    0.0
...                                                 ...    ...
2237  The TiON nanofilms were prepared by pulsed las...    1.0
2238  Samples were grown starting from commercially ...    1.0
2239  Circular pellets of diameter 10 mm were made a...    1.0
2240  The MSPLD system scheme is presented in Fig. 1...    1.0
2241  Commercially available targets of Al2O3, BaF2,...    1.0

[2242 rows x 2 columns]

In [343]:
df=merged_df

X_train, X_test, y_train, y_test = train_test_split(
    df['text'], df['label'], test_size=0.2, stratify=df['label'], random_state=143
)

vectorizer = TfidfVectorizer(max_features=10000, ngram_range=(1, 2))
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

model = LogisticRegression(class_weight='balanced', max_iter=1000)
model.fit(X_train_vec, y_train)
lr_model=model


y_pred = model.predict(X_test_vec)

print("Classification Report:")
print(classification_report(y_test, y_pred, digits=4))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

         0.0     0.9928    0.9835    0.9881       423
         1.0     0.7667    0.8846    0.8214        26

    accuracy                         0.9777       449
   macro avg     0.8798    0.9340    0.9048       449
weighted avg     0.9797    0.9777    0.9785       449

Confusion Matrix:
[[416   7]
 [  3  23]]


# Good Results on 0, Make sure to get on 1 as well
## Importance of classifying 1 is more than classifying 0 in this case

In [346]:
import pandas as pd

dff = pd.read_excel("paragraphsbefore2.xlsx")
dff = dff[['Paragraph', 'Label']][:1001]

dff = dff.rename(columns={
    'Paragraph': 'text',
    'Label': 'label'
})

dff

text  label
0     The SLs were grown by pulsed laser deposition ...    1.0
1     Thin films of w-BN are grown by pulsed laser d...    1.0
2     Interconnected ZnO core/shell nanorods were gr...    1.0
3     Before deposition, the substrates were heated ...    1.0
4     SrHA nanocrystals were synthesized in N2 atmos...    1.0
...                                                 ...    ...
996   The target for the pulsed laser deposition was...    1.0
997   The IMO thin films were deposited on quartz su...    1.0
998   A non-stoichiometric Bi1.1FeO3 target synthesi...    1.0
999   The experimental setup used to deposit the AlN...    1.0
1000  Target for the pulsed laser deposition was pre...    1.0

[1001 rows x 2 columns]

In [347]:
merged_df = pd.concat([df, dff], ignore_index=True)

merged_df

text  label
0     Today a variety of additive manufacturing (AM)...    0.0
1     The Ti-6Al-4V alloy, a popular α (hexagonal cl...    0.0
2     The microstructure transformation is another r...    0.0
3     For LMWD of Ti-6Al-4 V alloy, solutions for im...    0.0
4     Accordingly, this work aims to study the influ...    0.0
...                                                 ...    ...
3238  The target for the pulsed laser deposition was...    1.0
3239  The IMO thin films were deposited on quartz su...    1.0
3240  A non-stoichiometric Bi1.1FeO3 target synthesi...    1.0
3241  The experimental setup used to deposit the AlN...    1.0
3242  Target for the pulsed laser deposition was pre...    1.0

[3243 rows x 2 columns]

# TEST WITH 20% testcase
## Improved 1 as well as 0 class

In [348]:
df=merged_df

X_train, X_test, y_train, y_test = train_test_split(
    df['text'], df['label'], test_size=0.2, stratify=df['label'], random_state=143
)

vectorizer = TfidfVectorizer(max_features=10000, ngram_range=(1, 2))
X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

model = LogisticRegression(class_weight='balanced', max_iter=1000)
model.fit(X_train_vec, y_train)
lr_model=model


y_pred = model.predict(X_test_vec)

print("Classification Report:")
print(classification_report(y_test, y_pred, digits=4))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

         0.0     0.9764    0.9764    0.9764       423
         1.0     0.9558    0.9558    0.9558       226

    accuracy                         0.9692       649
   macro avg     0.9661    0.9661    0.9661       649
weighted avg     0.9692    0.9692    0.9692       649

Confusion Matrix:
[[413  10]
 [ 10 216]]


<div style="color: red; font-weight: bold; font-size: 3em;">
RESULTS:
</div>

Classification Report:

| Class         | Precision | Recall | F1-Score | Support |
|---------------|-----------|--------|----------|---------|
| 0.0           | 0.9764    | 0.9764 | 0.9764   | 423     |
| 1.0           | 0.9558    | 0.9558 | 0.9558   | 226     |
| **Accuracy**  |           |        | **0.9692**| 649     |
| Macro Avg     | 0.9661    | 0.9661 | 0.9661   | 649     |
| Weighted Avg  | 0.9692    | 0.9692 | 0.9692   | 649     |


Confusion Matrix:

|               | Predicted 0 | Predicted 1 |
|---------------|-------------|-------------|
| Actual 0      | 413         | 10          |
| Actual 1      | 10          | 216         |


In [350]:

import pandas as pd

df = pd.read_excel("para_data.xlsx")
df = df[['text', 'label']][2166:]

paragraphs = df['text'].tolist()
labels=df['label'].tolist()

paragraphs_vec = vectorizer.transform(paragraphs)

predictions = lr_model.predict(paragraphs_vec)

predictions_prob = lr_model.predict_proba(paragraphs_vec)[:, 1]

df_predictions = pd.DataFrame({
    'Paragraph': paragraphs,
    'Predicted Label': predictions,
    'Prediction':predictions_prob,
    'Original Label (by Amish)': labels
})

df_predictions.to_excel('predictions.xlsx', index=False)




